cantools/database/dataelement.py

```diff
@@ -0,0 +1,192 @@
+
+        #: ``True`` if the signal is signed, ``False`` otherwise. Ignore this
+        #: attribute if :data:`~cantools.db.Signal.is_float` is
+        #: ``True``.
+        self.is_signed: bool = is_signed
+
+        #: The unit of the signal as a string, or ``None`` if unavailable.
+        self.unit: Optional[str] = unit
+
+        #: Stores the raw start and end points of piecewise linear segments
+        #: empty for all other types
+        self.segment_boundaries_raw: PiecewiseSegment = []
+
+        #: Stores the scaled start and end points of piecewise linear segments
+        #: empty for all other types
+        self.segment_boundaries_scaled: PiecewiseSegment = []
+        if segment_boundaries is not None:
+            if not isinstance(scale, list) or not isinstance(offset, list):
+                raise ValueError(
+                    "Params scale and offset need to be of type list "
+                    "if segment boundaries are defined."
+                )
+
+            self._initialize_segment_limits(segment_boundaries)
+
+    def _initialize_segment_limits(self, segment_boundaries: PiecewiseSegment) -> None:
+        def convert(v, o, f):
+            return v * f + o
+
+        self.segment_boundaries: PiecewiseSegment = []
+        last_phys_max = None
+        for i, segment in enumerate(segment_boundaries):
+            self.segment_boundaries_raw.append(segment)
+            start, end = segment
+            params = [self._offset[i], self._scale[i]]  # type: ignore
+            scaled_segment = (convert(start, *params), convert(end, *params))
+            self.segment_boundaries_scaled.append(scaled_segment)
+            if last_phys_max is None:
+                last_phys_max = scaled_segment[1]
+                continue
+
+            if last_phys_max >= scaled_segment[0]:
+                logger.warning(
+                    f"Piecewise linear type: {self.name} has overlapping segments! "
+                    f"Segment {i} starts at phys val {scaled_segment[0]} "
+                    f"but one of the prev segments ended at {last_phys_max}. "
+                    "Encoding might be ambiguous."
+                )
+
+            last_phys_max = max(scaled_segment[1], last_phys_max)
+
+    def choice_string_to_number(self, string: str) -> int:
+        if self.choices is None:
+            raise ValueError(f"Signal {self.name} has no choices.")
+
+        for choice_number, choice_value in self.choices.items():
+            if str(choice_value) == str(string):
+                return choice_number
+
+        raise KeyError(f"Choice {string} not found in data element {self.name}.")
+
+    def _get_offset_scaling_from_list(self, val: float, segments: PiecewiseSegment) -> Tuple[float, float]:
+        for i, segment in enumerate(segments):
+            start, end = segment
+            if start <= val <= end:
+                return self._offset[i], self._scale[i]  # type: ignore
```

Comment:
why does the type need to be ignored here?

Reply:
`self._offset` and scale are of type type `Union[float, List[float]]` so mypy doesn't like it when I try to index them. However, in the `__init__()` I ensure that they are of type list if piecewise segments are defined so if we get to this line it would actually always be safe to assume that they are iterable.

Reply:
if you add `assert not isinstance(self._offset, float)` at the beginning of the function, `mypy` should be happy...